[SPARK-50216][SQL][TESTS] Update `CollationBenchmark` to invoke `collationNameToId` outside of cases

stevomitric opened the pull request:
### What changes were proposed in this pull request?
In this PR, UTF8_BINARY performance regression is addressed, that was first identified here https://github.com/apache/spark/pull/48721. The regression is traced back to this PR https://github.com/apache/spark/pull/48222 when it first occurred, however this isn't the actual source of performance degradation.


### Why are the changes needed?
The PR https://github.com/apache/spark/pull/48222 caused the regression because it changed the `collationNameToId` function and made it slightly slower by removing a short-circuit for fetching the UTF8_BINARY collation. However this function should be called fixed amount of times for each query and from the benchmark framework at most once - this was not the case and it was the largest contributor to performance regression.

This PR addresses the benchmarking framework to not call this function at each expression, but once per the test case.

### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Existing testing surface, benchmarks.


### Was this patch authored or co-authored using generative AI tooling?
No

stevomitric committed:
early-out check for utf8_binary

stevomitric committed:
fix benchmark

stevomitric committed:
revert changes

stevomitric committed:
add early-out check

MaxGekk reviewed (commented):
@stevomitric Could you regenerate results of the benchmark `CollationBenchmark`, please. We should expect better numbers after your changes, right?

stevomitric commented:
> @stevomitric Could you regenerate results of the benchmark CollationBenchmark, please. We should expect better numbers after your changes, right?

They are running, I will post them here once complete. We should expect same or better results.

MaxGekk commented on common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:
```diff
@@ -509,6 +509,10 @@ private CollationSpecUTF8(
       private static int collationNameToId(String originalName, String collationName)
           throws SparkException {
+        // Have a check for UTF8_BINARY collation to early-out and not introduce any regression.
+        if (UTF8_BINARY_COLLATION.collationName.equals(collationName)) {
```
You could avoid double comparison to `UTF8_BINARY_COLLATION` if you modify the `if` below like:
```java
        if (collationName.startsWith(UTF8_BINARY_COLLATION.collationName)) {
          if (collationName.length() == UTF8_BINARY_COLLATION.collationName.length()) {
            return UTF8_BINARY_COLLATION_ID;
          }
```

stevomitric replied:
Done. Added special case for `UTF8_LCASE` as well.

stevomitric replied:
FWI, according to the current (misleading) benchmarks, this approach yields ~40% slower results.

dongjoon-hyun commented:
Gentle ping, @stevomitric . Is the regenerated result ready?

Also, cc @panbingkun , FYI.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:
```diff
@@ -113,13 +113,13 @@ abstract class CollationBenchmarkBase extends BenchmarkBase {
       output = output)
     collationTypes.foreach { collationType => {
-      val collation = CollationFactory.fetchCollation(collationType)
+      val collationId = CollationFactory.collationNameToId(collationType)
```
Can we avoid touching the benchmark in the PR, @stevomitric ?

If we need this, we can proceed separately this `CollationBenchmark` change before your regression fix.

stevomitric replied:
The regression was partially due to benchmarks as well. The `CollationFactory.collationNameToId` function should only be called fixed amount of times per query, which shouldn't measurably impact the performance of the query.

So i think this is the right PR for it, since it was the leading contributor to the numbers seen in the results.

stevomitric committed:
add benchmarks

dongjoon-hyun reviewed (dismissed):
No, it's opposite. If that's true, we definitely should proceed it independently in order to measure how much they are affecting, @stevomitric . Please make a benchmark update PR first.

> The regression was partially due to benchmarks as well. The CollationFactory.collationNameToId function should only be called fixed amount of times per query, which shouldn't measurably impact the performance of the query.
> So i think this is the right PR for it, since it was the leading contributor to the numbers seen in the results.

panbingkun commented:
> No, it's opposite. If that's true, we definitely should proceed it independently in order to measure how much they are affecting, @stevomitric . Please make a benchmark update PR first.
> 
> > The regression was partially due to benchmarks as well. The CollationFactory.collationNameToId function should only be called fixed amount of times per query, which shouldn't measurably impact the performance of the query.
> > So i think this is the right PR for it, since it was the leading contributor to the numbers seen in the results.

Yes, I also agree, because in the previous PR (https://github.com/apache/spark/pull/48222) that may have generated performance regression, we did not see any modifications to `CollationBenchmark`. Therefore, in order to address this issue more clearly, can we do it separately? Thanks!

stefankandic reviewed (approved):
Great job finding the issue! Changes look good, but I also think it would make sense to do benchmark changes in a separate PR

stevomitric committed:
revert collation benchmark

stevomitric committed:
revert collation-factory changes, modify the benchmarks

stevomitric committed:
Merge branch 'master' into stevomitric/fix-utf8_binary-regression

stevomitric committed:
revert benchmarks

stevomitric committed:
revert changes

stevomitric commented:
> How about to update bechmark results after the revert https://github.com/apache/spark/commit/95b259ecc5c20649993e5d8173aa1b076c0c74ea

Decided to make this PR update the benchmarks, and not touch the `CollationFactory` class. Modifying the function in the way you proposed [here](https://github.com/apache/spark/pull/48804#discussion_r1834614134), resulted in very small perf benefit.

stevomitric committed:
add benchmark

dongjoon-hyun reviewed (approved):
Thank you for updating, @stefankandic . +1, LGTM.

> Decided to make this PR update the benchmarks, and not touch the CollationFactory class. Modifying the function in the way you proposed https://github.com/apache/spark/pull/48804#discussion_r1834614134, resulted in a very small perf benefit.

dongjoon-hyun commented:
Although I updated the PR title based on the AS-IS status, the PR description is still outdated. Could you make the PR description up-to-date with the AS-IS status, @stefankandic ?

stevomitric committed:
add non-ascii benchmarks

stevomitric commented:
> Could you make the PR description up-to-date with the AS-IS status?

Updated the description and added non-ascii collation benchmarks, @dongjoon-hyun .

MaxGekk reviewed (commented):
The failed GA [Run / Protobuf breaking change detection and Python CodeGen check](https://github.com/stevomitric/spark/actions/runs/11834590812/job/32981121006#logs) is not related to the changes, I believe.

+1, LGTM. Merging to master.
Thank you, @stevomitric and @dongjoon-hyun @HyukjinKwon @stefankandic for review.

dongjoon-hyun commented:
Thank you all!
